Add basic metric for Containers on Windows

KlwntSingh opened the pull request:
**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
Add pod level metric collection for Windows
    
    1. Changed receiver for awscontainerinsights to run for Windows with right
     metric collection sources.
    2. Added summary API in kubeletclient
    3. Add kubeletProvider to return metrics at different levels i.e. pod, contianer, node.
    4. Updated hostInfo providers to run for Windows.

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** <Describe what testing was performed and which tests were added.>
Tested locally and it worked

**Documentation:** <Describe the documentation added.>

tzifudzi commented on receiver/awscontainerinsightreceiver/internal/host/nodeCapacity.go:
```diff
@@ -49,11 +49,14 @@ func newNodeCapacity(logger *zap.Logger, options ...nodeCapacityOption) (nodeCap
-	if err := nc.osSetenv(goPSUtilProcDirEnv, hostProc); err != nil {
-		return nil, fmt.Errorf("NodeCapacity cannot set goPSUtilProcDirEnv to %s: %w", hostProc, err)
+	// On Windows, hostProc is empty
+	if hostProc != "" {
```
Nit: While the current approach effectively handles the condition, a way to potentially improve readability could be to introduce a variable like `isLinux`.

For example:
```
var isLinux = hostProc != ""
if isLinux {
     // Existing logic...
}
```

chadpatel replied:
I wonder if it would be better to have _win/_lin specific files for some of these implementation ones.  gopsutil does this
https://github.com/shirou/gopsutil/blob/master/process/process_windows.go

Another simple way but be to have the method call windows or linux sub-implementations within the same file.  Having them split out is nice because then you are not required to have the win/lin impls cross-platform compile

tzifudzi commented on receiver/awscontainerinsightreceiver/internal/host/utilconst_windows.go:
```diff
@@ -0,0 +1,14 @@
+package host
+
+// These variables are invalid for Windows
+const (
```
Another nit: This works, but could there be other more readable (and maintainable) ways to refactor this and enable the ability of having OS-specific behaviour or properties? Not sure how the codebase is what might work best so would rely on your good judgement.

KlwntSingh replied:
I saw similar pattern of defining OS specific const in repo.

tzifudzi commented on receiver/awscontainerinsightreceiver/internal/k8swindows/k8swindows.go:
```diff
@@ -0,0 +1,106 @@
+package k8swindows
+
```
nit: Can we ignore building this on Linux? 

```
//go:build windows
// +build windows
```

KlwntSingh replied:
Good point, addressed!

tzifudzi commented on receiver/awscontainerinsightreceiver/internal/k8swindows/k8swindows.go:
```diff
@@ -0,0 +1,106 @@
+		return result
+	}
+	metrics = k.decorateMetrics(metrics)
+	for _, cadvisorMetric := range metrics {
```
nit: Can the metric name in iteration be something like `k8s_summary_metric`?

KlwntSingh replied:
Addressed

tzifudzi commented on receiver/awscontainerinsightreceiver/internal/k8swindows/k8swindows.go:
```diff
@@ -0,0 +1,106 @@
+		tags[ci.AutoScalingGroupNameKey] = c.hostInfo.GetAutoScalingGroupName()
+
+		// add tags for EKS
+		tags[ci.ClusterNameKey] = c.hostInfo.GetClusterName()
```
nit: Seems like you covered the critical tags. Additional tags to consider if not already added. Might edit and add more as I think of more ideas
- platform e.g. Windows or Linux
- availability zone and region

KlwntSingh replied:
information about region is already part metrics internally, information about OS will be included in future PRs when refining metric labels

tzifudzi commented on receiver/awscontainerinsightreceiver/internal/k8swindows/kubelet.go:
```diff
@@ -0,0 +1,88 @@
+	hostIP := os.Getenv("HOST_IP")
+	kclient, err := kubeletutil.NewKubeletClient(hostIP, ci.KubeSecurePort, logger)
+	if err != nil {
+		return nil, errors.New("failed to initialize kubelet client")
```
To improve error handling and make debugging easier, can we please include the original error details so that a user will know what happened and how to rectify it?

KlwntSingh replied:
Addressed

tzifudzi commented on receiver/awscontainerinsightreceiver/internal/k8swindows/kubelet.go:
```diff
@@ -0,0 +1,88 @@
+}
+
+func new(logger *zap.Logger, info host.Info) (*kubeletSummaryProvider, error) {
+	hostIP := os.Getenv("HOST_IP")
```
Would it be helpful to check if `HOST_IP` was set correctly or is this checked later in the code?

Maybe we can add
```
if hostIP == "" {
  // return error or print warning if value is optional
}
```

KlwntSingh replied:
Before adding this validator, I want to verify a usecase of what happens when `HOST_IP` is not set. Based on results, i can add this additional check. if it fails, then i will add validation but it works then this is not required.

tzifudzi commented:
Does the project contain any unit/integration tests? Will these be added later?

tzifudzi commented on receiver/awscontainerinsightreceiver/internal/k8swindows/kubelet.go:
```diff
@@ -0,0 +1,88 @@
+
+	nodeCPUCores := k.hostInfo.GetNumCores()
+	for _, pod := range summary.Pods {
+		k.logger.Info(fmt.Sprintf("pod summary %v", pod.PodRef.Name))
```
Nit: Is there any way we can also show how many containers are running in the pod? If yes, and you agree it might be useful, lets add it. If not then nevermind.

KlwntSingh replied:
This information will be collected in separate metrics field but not exactly the number of containers rather cpu, memory metrics for each container in pod.

tzifudzi commented on receiver/awscontainerinsightreceiver/receiver.go:
```diff
@@ -197,7 +209,9 @@ func (acir *awsContainerInsightReceiver) Shutdown(context.Context) error {
 // collectData collects container stats from Amazon ECS Task Metadata Endpoint
 func (acir *awsContainerInsightReceiver) collectData(ctx context.Context) error {
 	var mds []pmetric.Metrics
-	if acir.cadvisor == nil && acir.k8sapiserver == nil {
```
nit: Similar to earlier feedback, maybe something like this might be more readable?

```
isLinux := acir.k8swindows != nil
isWindows := acir.cadvisor == nil && acir.k8sapiserver == nil && acir.k8swindows == nil
```

KlwntSingh replied:
I will address this in future PR since this complex logic might not be required.

KlwntSingh commented:
> Does the project contain any unit/integration tests? Will these be added later?

Yes, unit tests will be part of this repo but integ are part of different.
This PR is just to put in code structure. I will start adding unit tests from subsequent PRs

chadpatel commented:
you should branch off of aws-cwa-dev in the contributing repo and merge this to a new branch in that repo instead of merging it into your fork.  This will be helpful for having multiple developers and you can take advantage of the workflows and Amazon's github license or whatever for running those workflows

chadpatel commented on receiver/awscontainerinsightreceiver/internal/cadvisor/extractors/cpu_extractor.go:
```diff
@@ -33,7 +33,7 @@ func (c *CPUMetricExtractor) GetValue(info *cInfo.ContainerInfo, mInfo CPUMemInf
 	// When there is more than one stats point, always use the last one
 	curStats := GetStats(info)
-	metric := newCadvisorMetric(containerType, c.logger)
+	metric := NewCadvisorMetric(containerType, c.logger)
```
Intereseting! I wonder if we need to make the concept of a "cAdvisor metric" more generic, basically it is a container insights metric or container insights system metric.  Then that could sit in `internal` and be shared by both the cAdvisor and windows impl.

For now though I think it is fine to just do this, that is a decent refactor for not a ton of value other than better semantics

chadpatel commented on receiver/awscontainerinsightreceiver/internal/cadvisor/extractors/extractor.go:
```diff
@@ -44,7 +44,8 @@ type CAdvisorMetric struct {
 }
 
-func newCadvisorMetric(mType string, logger *zap.Logger) *CAdvisorMetric {
+// NewCadvisorMetric is public because used in k8swindows package, outside cadvisor package.
```
Nit - I would delete this comment, this is useful in a diff but I don't think it is useful for this comment to live on forever.  Packages/methods don't need to detail why they are public or private or whatever

KlwntSingh replied:
Addressed, removed the comment!

chadpatel commented on receiver/awscontainerinsightreceiver/internal/host/nodeCapacity.go:
```diff
@@ -49,11 +49,14 @@ func newNodeCapacity(logger *zap.Logger, options ...nodeCapacityOption) (nodeCap
-	}
-	if err := nc.osSetenv(goPSUtilProcDirEnv, hostProc); err != nil {
-		return nil, fmt.Errorf("NodeCapacity cannot set goPSUtilProcDirEnv to %s: %w", hostProc, err)
+	// On Windows, hostProc is empty
```
nit - I don't love this comment.  The check is for hostProc, not windows/linux.  The code should be evident that it needs a hostProc, all the reasons why it may not have a valid hostproc (i.e. windows) is not relevant

KlwntSingh replied:
Addressed.

chadpatel commented on receiver/awscontainerinsightreceiver/internal/stores/kubeletutil/kubeletclient.go:
```diff
@@ -9,6 +9,7 @@ import (
 	"go.uber.org/zap"
 	corev1 "k8s.io/api/core/v1"
+	stats "k8s.io/kubelet/pkg/apis/stats/v1alpha1"
 
```
any idea on the govuln check?

chadpatel replied:
is it something about being alpha?

chadpatel commented on receiver/awscontainerinsightreceiver/receiver.go:
```diff
@@ -74,26 +77,35 @@ func (acir *awsContainerInsightReceiver) Start(ctx context.Context, host compone
-		if err != nil {
-			return err
-		}
+		// On Windows, K8sWindows provider is used which internally uses kubelet summary API for metrics.
```
IMO this is an unnecessary comment

KlwntSingh replied:
Addressed

chadpatel commented on receiver/awscontainerinsightreceiver/receiver.go:
```diff
@@ -197,7 +209,9 @@ func (acir *awsContainerInsightReceiver) Shutdown(context.Context) error {
 	var mds []pmetric.Metrics
-	if acir.cadvisor == nil && acir.k8sapiserver == nil {
+	// On linux, acir.k8swindows will be nil.
+	// On Windows, both acir.cadvisor and acir.k8sapiserver will be nil.
```
I don't think this is a useful comment, it is basically documenting your assumption about the entry criteria

KlwntSingh replied:
addressed.

chadpatel commented on receiver/awscontainerinsightreceiver/receiver.go:
```diff
@@ -39,6 +41,7 @@ type awsContainerInsightReceiver struct {
 	cancel            context.CancelFunc
 	cadvisor          metricsProvider
 	k8sapiserver      metricsProvider
+	k8swindows        metricsProvider
```
my two cents - k8swindows and cadvisor are **the same**, no reason to not just populate cadvisor with two different impls, one impl for windows and one impl for linux (actually using cadvisor).

Then you don't need special code in the receiver for k8swindows.GetMetrics()

I am open to renaming it, kubeletProvider or something

KlwntSingh replied:
Addressed.
I have removed `k8swindows` from this struct. Renamed `cadvisor` field to  `containerMetricProvider` and reused it for both cadvisor and k8sWindows implementation.

chadpatel commented:
General comment, your implementation for windows CI is really an alternate implementation to the cAdvisor code.  cAdvisor as a name is probably over-used, we could rename to kubelet or kubeletStats or kubeletProvider or nodeStats or whatever but I would avoid creating a whole new "windows" thing when you essentially just need to create an alternate implementation of the existing cAdvisor "provider"

This will reduce the code sprawl and minimize the number of windows/linux checks you have to make

KlwntSingh committed:
Add pod level metric collection for Windows

This PR defines code structure for metric provider which works on Windows.

### Changelog
1. Changed receiver.go in awscontainerinsights to run for Windows with metric provider.
2. Added summary API in kubeletclient
3. Add kubeletProvider to return metrics at different levels i.e. pod, contianer, node.
4. Updated hostInfo providers to run for Windows.
5. Updated ebsVolume Info provider to run for Windows.

## Todos:
1. Define correct ebsVolume Info provider for Windows
2. Change logic around k8s leader election to run for Windows

KlwntSingh commented:
> General comment, your implementation for windows CI is really an alternate implementation to the cAdvisor code. cAdvisor as a name is probably over-used, we could rename to kubelet or kubeletStats or kubeletProvider or nodeStats or whatever but I would avoid creating a whole new "windows" thing when you essentially just need to create an alternate implementation of the existing cAdvisor "provider"
> 
> This will reduce the code sprawl and minimize the number of windows/linux checks you have to make

KlwntSingh committed:
Fix workflow issues

KlwntSingh committed:
fix PR checks

KlwntSingh committed:
Fix go.mod in cmd/configschema
